adaptdl/adaptdl/speedup.py

```diff
@@ -171,31 +194,78 @@ def __call__(self, nodes, replicas, return_local_bsz=False):
         self._mem_speedup[mem_indices] = speedup[ret_indices]
         self._mem_local_bsz[mem_indices] = local_bsz[ret_indices]
 
+        ret_atomic_bsz, ret_accumulation_steps = \
+            self._partition_local_bsz(ret_local_bsz, replicas)
+
         if isscalar:
             ret_speedup = ret_speedup.item()
-            ret_local_bsz = ret_local_bsz.item()
-
-        return ((ret_speedup, ret_local_bsz)
-                if return_local_bsz else ret_speedup)
+            ret_atomic_bsz = int(ret_atomic_bsz.item())
+            ret_accumulation_steps = int(ret_accumulation_steps.item())
+        if return_config:
+            return (ret_speedup, (ret_atomic_bsz, ret_accumulation_steps))
+        else:
+            return ret_speedup
+
+    def _partition_local_bsz(self, local_bsz, replicas):
+        if (self._max_local_bsz is not None and self._accumulation and (
+                np.any(self._max_local_bsz < local_bsz)
+                or np.any(replicas == 1))):
+            accumulation_steps = np.maximum(
+                np.ceil(local_bsz / self._max_local_bsz),
+                1)
+            atomic_bsz = np.minimum(
+                self._max_local_bsz, np.ceil(local_bsz / accumulation_steps))
+
+            # Need to prevent a single replica from scaling up the
+            # batch size without using gradient accumulation
+            single_replica_predicate = replicas == 1
+            single_step_predicate = np.logical_or(
+                local_bsz == self._init_batch_size,
+                local_bsz <= 2 * self._min_local_bsz)
```

Comment:
Is the predicate just `(replicas == 1) & (accumulation_steps <= 1)`?

Reply:
I think that would result in slightly different behavior. With `(replicas == 1) & (accumulation_steps <= 1)`, then that would result in batch sizes in between `self._init_batch_size` and `max_local_bsz` being divided up into `(self._init_batch_size, 1)`, instead of `(batch_size/2, 2)`. I think the second is preferable.